atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/TransactionCacheValueStoreImpl.java

```diff
@@ -130,23 +134,49 @@ public TransactionCacheValueStore createWithFilteredSnapshot(CommitUpdate commit
                 .collect(Collectors.toSet());
     }
 
+    private void cacheHitInternal(TableReference table, Cell cell, CacheValue value) {
+        localUpdates.compute(CellReference.of(table, cell), (_unused, previousValue) -> {
+            if (previousValue != null) {
+                throw new SafeIllegalStateException(
+                        "Should not be attempting to record hits for keys that have been written to or read from the"
+                                + " KVS",
+                        UnsafeArg.of("table", table),
+                        UnsafeArg.of("cell", cell),
+                        UnsafeArg.of("previousValue", previousValue));
+            } else {
+                return LocalCacheEntry.hit(value);
+            }
+        });
+    }
+
     private Map<Cell, CacheValue> getLocallyCachedValues(TableReference table, Set<Cell> cells) {
         return KeyedStream.of(cells)
                 .map(cell -> localUpdates.get(CellReference.of(table, cell)))
                 .filter(Objects::nonNull)
+                .filter(value -> !value.status().equals(Status.WRITE))
                 .map(LocalCacheEntry::value)
+                .map(TransactionCacheValueStoreImpl::getReadValue)
                 .collectToMap();
     }
 
-    private void cacheRemoteWriteInternal(CellReference cellReference, CacheValue value) {
-        if (snapshot.isWatched(cellReference.tableRef()) && snapshot.isUnlocked(cellReference)) {
-            localUpdates.put(cellReference, LocalCacheEntry.write(value));
+    private Set<Cell> getLocallyWrittenCells(TableReference table, Set<Cell> cells) {
+        return KeyedStream.of(cells)
+                .map(cell -> localUpdates.get(CellReference.of(table, cell)))
+                .filter(Objects::nonNull)
+                .filter(value -> value.status().equals(Status.WRITE))
+                .keys()
+                .collect(Collectors.toSet());
+    }
+
+    private void recordRemoteWriteInternal(CellReference cellReference) {
+        if (snapshot.isUnlocked(cellReference)) {
```

Comment:
Just to verify I understand this correctly: the current transaction already caches the write anyway, so this check is only for the benefit of the read transaction for serialisable conflict checking?

Reply:
The `isUnlocked` also includes whether this table is watched or not, so checking this does make sure that we don't accidentally register information for non-watched tables (which just saves memory, really).